### PR TITLE
Add host header support for JSON

### PR DESCRIPTION
**Pull request by vjeffz**
Add host header support for JSON, see issue: #727 

```
python gsutil.py -DD -o "Credentials:gs_json_host_header=www.googleapis.com" -o "Credentials:gs_json_host=restricted.googleapis.com" ls gs://
```

### CODE REVIEW

**Comment by vjeffz** [comment]
I signed it!

**Comment by vjeffz** [comment]
> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project (if not, look below for help). Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please visit https://cla.developers.google.com/ to sign.**
> 
> Once you've signed (or fixed any issues), please reply here (e.g. `I signed it!`) and we'll verify it.
> 
> #### What to do if you already signed the CLA
> ##### Individual signers
> * It's possible we don't have your GitHub username or you're using a different email address on your commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
> ##### Corporate signers
> * Your company has a Point of Contact who decides which employees are authorized to participate. Ask your POC to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the Google project maintainer to [go/cla#troubleshoot](http://go/cla#troubleshoot) ([Public version](https://opensource.google.com/docs/cla/#troubleshoot)).
> * The email used to register you as an authorized contributor must be the email used for the Git commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * The email used to register you as an authorized contributor must also be [attached to your GitHub account](https://github.com/settings/emails).
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2FGoogleCloudPlatform%2Fgsutil%2Fpull%2F728) for more info**.

I signed it!

**Comment by houglum** [comment]
For the CLA check, you'll need to sign it with the email address you used to author your commits.  It's still not finding that one in the CLA list.

File: gslib/gcs_json_api.py [hunk 1]
```diff
@@ -223,7 +223,13 @@ def __init__(self, bucket_storage_uri_class, logger, status_queue,
 
     self.global_params = apitools_messages.StandardQueryParameters(
         trace='token:%s' % trace_token) if trace_token else None
+
     additional_http_headers = {}
+
+    gs_json_host_header = config.get('Credentials', 'gs_json_host_header', None)
```
Comment (houglum) [inline gslib/gcs_json_api.py]: I was going to suggest also adding this for `gs_host`, but it turns out that Boto already supports this:
https://github.com/gsutil-mirrors/boto/blob/91ba037e54ef521c379263b0ac769c66182527d7/boto/provider.py#L215

So no action needed. Mostly just leaving this comment here as a reminder to my future self :)

**Comment by vjeffz-srcd** [comment]
I've signed it

**Comment by houglum** [comment]
Our CLA list apparently doesn't have your GitHub username tied to the email address you used to author the commits, but I did see that the email address used now registers as having a CLA, so I'll mark this as `cla: yes`.

**Comment by vjeffz** [comment]
> Could you also add these headers (`gs_json_host_header`, `<provider>_host_header`) to the list of options generated in user boto config files? This includes the list here:
> [gsutil/gslib/commands/config.py](https://github.com/GoogleCloudPlatform/gsutil/blob/d7281accc526df971d0c4c66327776b5dd1299f9/gslib/commands/config.py#L199)
> 
> Line 199 in [d7281ac](/GoogleCloudPlatform/gsutil/commit/d7281accc526df971d0c4c66327776b5dd1299f9)
> 
>        gs_host 
> as well as the text here:
> [gsutil/gslib/commands/config.py](https://github.com/GoogleCloudPlatform/gsutil/blob/d7281accc526df971d0c4c66327776b5dd1299f9/gslib/commands/config.py#L1070)
> 
> Line 1070 in [d7281ac](/GoogleCloudPlatform/gsutil/commit/d7281accc526df971d0c4c66327776b5dd1299f9)
> 
>  '#%s_host = <alternate storage host address>\n'

Done

**Comment by houglum** [comment]
Thanks! I added a couple commits to add comments for the `gs_host_header` in addition to the `gs_json_host_header`. Will merge shortly.

**Pull request merged**